### PR TITLE
PrettyPrinter: init (WIP)

### PR DESCRIPTION
**Pull request by dtzSiFive**
Introduces PrettyPrinter for use emitting "intelligently" formatted output given a stream of tokens (strings, breaks, begin/end marking groups).

Based on Oppen's algorithm (citation in PrettyPrinter.cpp, as well as some notes/attributions).

See the file headers for more information, and the unit test for a few function call examples at different widths.  Output style depends on strategy used for groups and break placement/parameters, these are not meant to be necessarily how we output anything but demonstrate and test a few possibilities.

This PR changes no CIRCT behavior.

Draft to expose this while under development.

Contains commits from #4082 to enable unit testing again.

Plan:
* [ ] Add a builder/convenience helper next to make using this much easier to use and to manage the Token's while they're streaming through the printer.
* [ ] Start integrating this with ExportVerilog.

### CODE REVIEW

File: include/circt/Support/PrettyPrinter.h [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+  };
+
+private:
+  union {
+    StringInfo stringInfo;
+    BreakInfo breakInfo;
+    BeginInfo beginInfo;
+    EndInfo endInfo;
+  } data;
+  const Kind kind;
```
[the remark this answers is not in the document]
Comment (dtzSiFive) [reply include/circt/Support/PrettyPrinter.h:65]: Thanks for bringing this up!

I couldn't find something, somewhat surprisingly (alternative/suggestions welcome)!  Closest AFAICT is: https://reviews.llvm.org/D98477 which never landed.  There's `PointerSumType` but these Token's are small and ideally could be stored on the stack/contiguously (`std::vector<Token>`).

You're right re:LLVM RTTI which is familiar but may not be serving us particularly well here, and don't expect clients to be juggling these themselves much (which might motivate using what's LLVM-y) but we'll see.

`std::variant` (probably just `using Token = std::variant<StringToken,...>;`?) would be a lot neater and has nice guarantees, but was worried about some unexpected performance issues (esp if using `std::visit`) but should circle back and compare against such alternative implementations.  Discriminated union being comparatively simple to reason about re:layout+generated code was the idea, but if that seems problematic I'm happy to switch to `std::variant` until there's reason to not use it.

**Comment by teqdruid** [comment]
I'm curious how much this improves the Verilog output. Can you show a before and after here?

**Review by uenoku — COMMENTED** [review]
Thank you for starting PrettifyPrinter! I'm looking forward it will be integrated into ExportVerilog! Just out of curiosity, is there any chance that we can use clang-format as a library of PrettifyFormating?

File: unittests/Support/PrettyPrinterTest.cpp [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+              int a2, int b2, int a3, int b3,
+              int a4, int b4, int a5, int b5,
+              int a6, int b6, int a7, int b7,
+              float xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+              ),
+      int a4, int b4, int a5, int b5, int a6,
+      int b6, int a7, int b7,
+      float xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+      );
+)"""));
```
[the remark this answers is not in the document]
Comment (dtzSiFive) [reply unittests/Support/PrettyPrinterTest.cpp:269]: :grin:, great! The token stream there is a bit messy, you can find some simpler examples (leveraging the convenience classes a bunch which will more closely match how we emit this) here if you're curious: https://github.com/dtzSiFive/circt/blob/ffd1b7c6a0a2961f9440349c81207bdae07a14c6/unittests/Support/PrettyPrinterTest.cpp#L427 (and to end of the file).

**Comment by dtzSiFive** [comment]
> Thank you for starting PrettifyPrinter! I'm looking forward it will be integrated into ExportVerilog! Just out of curiosity, is there any chance that we can use clang-format as a library of PrettifyFormating?

This probably would work with some effort, and would of course save us from building our own bits and allow us to leverage their great work and optimizations to a hard problem!

This didn't seem like a good path when I was looking at it (wrong interface, expensive algorithm somewhat unpredictably, possibly lots of work to adapt more completely to verilog's constructs, unclear if that'd be superficial or more deeply involved).  But not being super familiar with its internals, there definitely might be ways to leverage their hard work and battle-tested (on other languages) implementation and optimizations (if not through its interface, than with a modified version at least in the short term).  This is probably worth investigating further if someone's got cycles/interest.  I'd be willing/interested in sitting down and taking a look at this with you if you'd like to pursue this!  There's a lot of benefit to be had if this can be made suitable for our needs!

## clang-format support for verilog

I think I've mentioned this before, but clang-format actually does have basic support for verilog and recently-ish (last few months IIRC) a number of patches for various verilog constructs were added, but in my testing it doesn't work very well.  Presumably it could be improved upon, and probably will (someone seems to be on that path?), but I'm not sure how hard it'd be to get it working with all the constructs we care about (and how compatible such would be with its internals/optimizations).

## libformat ?

I don't think libFormat as-is is suitable for throwing AST/tokens/etc at, it seems to currently have a very slim simple interface: https://clang.llvm.org/docs/LibFormat.html .
It's possible we could integrate with it more directly ourselves, to drive it more directly and avoid having it re-lex/parse an output stream...?

## clang-format performance on verilog, an anecdote

Anecdotally, FWIW it choked (killed after few hours) on a 16MB file I fed it, but unsure if that was somehow a particular pathological case or if it just doesn't scale up especially well (obviously we'll want to format much larger designs).  Breaking things apart may help, but it originally at least operates on "UnwrappedLine"s so it shouldn't be, say, solving for the ideal format of an entire module or file at once.  It doesn't have that sort of performance issue when pushing lots of C/C++ through it, so there's hope :smile:.

## algorithmic suitability

Its fundamental approach may be a bit too expensive (dynamic programming-based searching state space of format possibilities to find lowest "cost" through graph of decision points (I think), the details of their approach are not documented AFAICT but there are some resources elsewhere that describe their approach in the abstract) to be on-by-default during the emission phase, was one of my main concerns.  Tweaking available parameters (?) limiting states searched, for example, may make it more suitable, and now that we limit expression sizes somewhat maybe it'd work better, but it's still solving a "hard"-er problem than we really need (not a problem if it's fast enough!).  IIRC similar tools have had success with alternative solving strategies for situations where you don't need "optimal" just the first reasonable-ish result.

## other tools / languages

The `rustfmt` tool is similar in approach (AFAIK) to `clang-format`, but it too can't handle a bunch of things (dunno if due to rust language complications, or what-- see `prettyplease` docs for examples/information, as they motivated its creation, basically it just punts sometimes).  As a result the rust compiler has an implementation of the same algorithm here (Oppen), `rustc_ast_pretty`-- which `prettyplease` is based on and most of its improvements have been upstreamed to `rustc_ast_pretty` by the author, although they use the underlying algorithm/primitives differently for a few reasons.  I mention both of these in the header of `PrettyPrinter.cpp` as they were very helpful to study when putting this together as for the most part they stick to the paper's algorithm (be careful with some of the alterations).  `swift-format` uses an Oppen-based approach as well, FWIW (with more changes to accommodate Swift language + style preferences-- they document this very well!).  All of these are useful for informing how to use Oppen-based pretty-printer primitives to nicely print language constructs.  Anyway my point for all that is that if they found it useful to have a simpler approach (for generated code especially), well, we're in good company :smile:.

 `dartfmt` has lots of goodness to learn from, and the author wrote a great blog post about these lessons, if you're interested.

re:verilog formatters-- presently I've had better luck with the verible formatter (but didn't work great in my limited testing, and it currently has a flag to enable trying to wrap lines which is off by default, and it seems for good reason-- I couldn't increase the search space enough to run my test cases through it).  Building something based on slang (`slang-format`!) has potential as well :smile: !

## functional formulations, research literature

There a lots of work on functional approaches/formulations of this problem, all of which are based (sometimes transitively, although many explicitly mention this) on Oppen-- I can share a list if you're curious, I've been having fun going through them (the thesis introducing `scalafmt` should be a good starting point for walking the reference graph forwards/backwards, also [Elixir](https://hexdocs.pm/elixir/Inspect.Algebra.html) and [Wadler](https://homepages.inf.ed.ac.uk/wadler/papers/prettier/prettier.pdf))-- which are very (!) compelling for their elegance (layout combinators? yes please ([`rfmt`, for R](https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/44667.pdf))) but didn't seem reasonable to map down to C++/our usage and in the end seem to all still require a long-tail of special casing and optimizations to make them perform efficiently (as they report).

Honestly I was originally planning to go in that sort of direction.

## summary

Regardless, this is not at all meant to be a strong rebuttal of the idea, just sharing what I've been looking through-- I still can't really shake "maybe we could use clang-format" :upside_down_face: entirely, especially because it works so very well, we already use it constantly, and is part of the LLVM project already!  And how great would it be if its magic worked on verilog too?!

In short, not sure, probably could be made to work!  It's not an entirely incompatible approach-- we could have a fast printer that does a reasonable job given its limited lookahead / linear costs and optionally pipe the output through a more powerful tool (integrating as much as makes sense) when requested.

**Maybe we could reach out to clang-format folks and bounce the idea off them** ?

**Comment by dtzSiFive** [comment]
> I'm curious how much this improves the Verilog output. Can you show a before and after here?

Meant to respond to this earlier, whoops, glad you're curious!  I absolutely intend to show before/after examples when that comes together, but don't have it integrated to emission properly yet (this and the other PR introduce supporting infra).  What we do presently is really basic (when emitting an expression within a statement, if the expression is longer than the configurable length parameter we break on spaces and wrap at `currentBaseIndentOfStatement * 2` or thereabouts), so any sort of breaking at reasonable points + multiple indentation levels when wrapping should be a significant improvement, even if not "optimal".  I'm eager to see this on real examples too, proof is in the demo-pudding, stay tuned! :+1:
